File: javalang/quicksort/QuickSort.java
package javalang.quicksort;

import java.util.Arrays;

public class QuickSort {
    public static void quickSort(int[] array) {
        quickSort(array, 0, array.length - 1);
    }

    private static void quickSort(int[] array, int first, int last) {
        if (last > first) {
            int pivotIndex = partition(array, first, last);
            quickSort(array, first, pivotIndex - 1);
            quickSort(array, pivotIndex + 1, last);
        }
    }

    private static int partition(int[] array, int first, int last) {
        int pivot = array[first];
        int low = first + 1;
        int high = last;

        while (high > low) {
            while (low <= high && array[low] <= pivot) {
                low++;
            }

            while (low <= high && array[high] > pivot) {
                high--;
            }

            if (high > low) {
                int temp = array[high];
                array[high] = array[low];
                array[low] = temp;
            }
        }

        while (high > first && array[high] >= pivot) {
            high--;
        }

        if (pivot > array[high]) {
            array[first] = array[high];
            array[high] = pivot;
            return high;
        } else {
            return first;
        }
    }

    public static void main(String[] args) {
        System.out.println("testing quick sort in java...");
        int[] unsortedArray = { 9, 5, 7, 1, 6, 2, 3, 8, 4 };

        System.out.println("unsorted array: " + Arrays.toString(unsortedArray));
        QuickSort.quickSort(unsortedArray);
        System.out.println("sorted array: " + Arrays.toString(unsortedArray));
    }
}

File: rust/algorithms_in/src/quick_sort.rs
pub fn quick_sort(array: &mut [usize], first: usize, last: usize) {
    if last > first {
        let pivot_index: usize = partition(array, first, last);
        if pivot_index > 0 {
            quick_sort(array, first, pivot_index - 1);
        }
        quick_sort(array, pivot_index + 1, last);
    }
}

fn partition(array: &mut [usize], first: usize, last: usize) -> usize {
    let pivot: usize = array[first];
    let mut low: usize = first + 1;
    let mut high: usize = last;

    while high > low {
        while low <= high && array[low] <= pivot {
            low += 1;
        }

        while low <= high && array[high] > pivot {
            high -= 1;
        }

        if high > low {
            let temp: usize = array[high];
            array[high] = array[low];
            array[low] = temp;
        }
    }

    while high > first && array[high] >= pivot {
        high -= 1;
    }

    if pivot > array[high] {
        array[first] = array[high];
        array[high] = pivot;
        return high;
    } else {
        return first;
    }
}